src/compression/mod.rs

```diff
@@ -408,16 +408,81 @@ fn mod_p(x: i32, y: i32) -> i32 {
 mod optimize_bytes {
 
     /// Integrate over all differences to the previous value in order to reconstruct sample values.
-    pub fn differences_to_samples(buffer: &mut [u8]){
-        for index in 1..buffer.len() {
-            buffer[index] = (buffer[index - 1] as i32 + buffer[index] as i32 - 128) as u8; // index unsafe but handled with care and unit-tested
+    pub fn differences_to_samples(buffer: &mut [u8]) {
+        // The naive implementation is very simple:
+        //
+        // for index in 1..buffer.len() {
+        //    buffer[index] = (buffer[index - 1] as i32 + buffer[index] as i32 - 128) as u8;
+        // }
+        //
+        // But we process elements in pairs to take advantage of instruction-level parallelism.
+        // When computations within a pair do not depend on each other, they can be processed in parallel.
+        // Since this function is responsible for a very large chunk of execution time,
+        // this tweak alone improves decoding performance of RLE images by 20%.
+        if let Some(first) = buffer.get(0) {
+            let mut previous = *first as i16;
+            for chunk in &mut buffer[1..].chunks_exact_mut(2) {
+                // no bounds checks here due to indices and chunk size being constant
+                let diff0 = chunk[0] as i16;
+                let diff1 = chunk[1] as i16;
+                // these two computations do not depend on each other, unlike in the naive version,
+                // so they can be executed by the CPU in parallel via instruction-level parallelism
+                let sample0 = (previous + diff0 - 128) as u8;
+                let sample1 = (previous + diff0 + diff1 - 128 * 2) as u8;
+                chunk[0] = sample0;
+                chunk[1] = sample1;
+                previous = sample1 as i16;
+            }
+            // handle the remaining element at the end not processed by the loop over pairs, if present
+            for elem in &mut buffer[1..].chunks_exact_mut(2).into_remainder().iter_mut() {
+                let sample = (previous + *elem as i16 - 128) as u8;
+                *elem = sample;
+                previous = sample as i16;
+            }
         }
     }
 
     /// Derive over all values in order to produce differences to the previous value.
     pub fn samples_to_differences(buffer: &mut [u8]){
-        for index in (1..buffer.len()).rev() {
-            buffer[index] = (buffer[index] as i32 - buffer[index - 1] as i32 + 128) as u8; // index unsafe but handled with care and unit-tested
+        // naive version:
+        // for index in (1..buffer.len()).rev() {
```

Comment:
love that you added a comment with the core idea of the algorithm :)

Reply:
Yes. My big gripe with lots of C code is that it only includes the optimized algorithm without explaining what it actually does. I really wanted to avoid this here.